client/java/src/main/java/com/linecorp/centraldogma/client/MappingWatcher.java

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright 2019 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.centraldogma.client;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static java.util.Objects.requireNonNull;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.function.BiConsumer;
+import java.util.function.Function;
+
+import javax.annotation.Nullable;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.collect.Maps;
+
+import com.linecorp.centraldogma.common.Revision;
+
+final class MappingWatcher<T, U> implements Watcher<U> {
+
+    private static final Logger logger = LoggerFactory.getLogger(MappingWatcher.class);
+
+    static <T, U> MappingWatcher<T, U> of(Watcher<T> parent, Function<? super T, ? extends U> mapper,
+                                          Executor executor, boolean closeParentWhenClosing) {
+        requireNonNull(parent, "parent");
+        requireNonNull(mapper, "mapper");
+        requireNonNull(executor, "executor");
+        // TODO(minwoo): extract mapper function and combine it with the new mapper.
+        return new MappingWatcher<>(parent, mapper, executor, closeParentWhenClosing);
+    }
+
+    private final Watcher<T> parent;
+    private final Function<? super T, ? extends U> mapper;
+    private final boolean closeParentWhenClosing;
+    private final CompletableFuture<Latest<U>> initialValueFuture = new CompletableFuture<>();
```

Comment:
Q) I wonder if we can't just use `parent.initialValueFuture`.   I think it's possible that users use a pattern like the following where intermediate mapped/child watchers aren't closed.

```
final Watcher<String> watcher = dogma.client()
                                      .forRepo("foo", "bar")
                                      .watch(Query.ofText("/baz.txt")).forever();
// the mapped watcher won't be closed
Consumer<String> consumer = c -> { /* do something */ };
watcher.map(a -> {
    consumer.accept(a);
    return null;
});
watcher.close();
```

I'm wondering if this may be a potential source of memory leak (if the client uses `initialValueFuture` do to something)

Reply:
>  I wonder if we can't just use parent.initialValueFuture

The complete value of the future will be different from the parent one so I think we cannot use it.
If the mapper returns `null`, a watcher currently leaves a warning log here.
https://github.com/line/centraldogma/pull/651/files#diff-5b249ef74ec0481e49cfd183e95e9629d59dd78ee63f424b3bdee6ed00a5c7daR79

If that's reasonable to return null in the mapper, I think we can remove the warning log and complete the future with the `null`. Let me do it in that way. 😉